File: jump_game.rb
def can_jump(nums)
  jump_game_helper(nums, nums.length - 1, {})
end

def jump_game_helper(nums, current_index)
  return true if current_index == 0 ||
                 memo[current_index] = true

  (0...current_index).each do |inner_index|
    return true if nums[inner_index] >= (current_index - inner_index) &&
                   jump_game_helper(nums, inner_index)
  end

  false
end

File: dynamic_programming/jump_game.js
// Given an array of non-negative integers, you are initially positioned at the first index of the array.
// Each element in the array represents your maximum jump length at that position.
// Determine if you are able to reach the last index.
// For example:
// A = [2,3,1,1,4], return true.
// A = [3,2,1,0,4], return false.
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canJump = function(nums) {
  const memo = [];
  return helper(nums, nums.length - 1, memo);
}

function helper(nums, current_index, memo) {
  if (current_index === 0) {
    return true;
  }

  for (let i = 0; i < current_index; i += 1) {
    if (i + nums[i] >= current_index) {
      if (memo[i] || helper(nums, i, memo)) {
        memo[i] = true;
        return true;
      } else {
        memo[current_index] = false;
      }
    }
  }

  return false;
}
